cmd/collectors/restperf/plugins/volumetopclients/volumetopclients.go

```diff
@@ -0,0 +1,409 @@
+		return nil
+	}
+
+	topClients, err := v.fetchTopClients(volumes.Values(), svms.Values(), metric)
+	if err != nil {
+		return err
+	}
+
+	mat := v.data[matrixName]
+	if mat == nil {
+		return nil
+	}
+	for i, client := range topClients {
+		clientIP := client.Get("client_ip").String()
+		vol := client.Get("volume.name").String()
+		svm := client.Get("svm.name").String()
+		value := client.Get(metric).Float()
+		instance, err := mat.NewInstance(strconv.Itoa(i))
+		if err != nil {
+			v.Logger.Warn().Str("volume", vol).Msg("error while creating instance")
+			continue
+		}
+		instance.SetLabel("volume", vol)
+		instance.SetLabel("svm", svm)
+		instance.SetLabel("client_ip", clientIP)
+		setMetricFunc(mat, instance, value)
+	}
+	return nil
+}
+
+func (v *Volume) setOpsMetric(mat *matrix.Matrix, instance *matrix.Instance, value float64) {
+	var err error
+	m := mat.GetMetric("ops")
+	if m == nil {
+		if m, err = mat.NewMetricFloat64("ops"); err != nil {
+			v.Logger.Warn().Err(err).Str("key", "ops").Msg("error while creating metric")
+			return
+		}
+	}
+	if err = m.SetValueFloat64(instance, value); err != nil {
+		v.Logger.Error().Err(err).Str("metric", "ops").Msg("Unable to set value on metric")
+	}
+}
+
+func (v *Volume) setDataMetric(mat *matrix.Matrix, instance *matrix.Instance, value float64) {
+	var err error
+	m := mat.GetMetric("data")
+	if m == nil {
+		if m, err = mat.NewMetricFloat64("data"); err != nil {
+			v.Logger.Warn().Err(err).Str("key", "data").Msg("error while creating metric")
+			return
+		}
+	}
+	if err = m.SetValueFloat64(instance, value); err != nil {
+		v.Logger.Error().Err(err).Str("metric", "data").Msg("Unable to set value on metric")
+	}
+}
+
+func (v *Volume) fetchVolumesWithActivityTrackingEnabled() (*set.Set, error) {
+	var (
+		result []gjson.Result
+		err    error
+	)
+	if v.volumeInterface != nil {
+		return v.volumeInterface.fetchVolumesWithActivityTrackingEnabled()
+	}
+	va := set.New()
+	query := "api/storage/volumes"
+	href := rest.NewHrefBuilder().
+		APIPath(query).
+		Fields([]string{"svm.name", "name"}).
+		Filter([]string{"activity_tracking.state=on"}).
+		Build()
+
+	if result, err = collectors.InvokeRestCall(v.client, href, v.Logger); err != nil {
+		return va, err
+	}
+
+	for _, volume := range result {
+		volName := volume.Get("name").String()
+		svmName := volume.Get("svm.name").String()
+		va.Add(svmName + keyToken + volName)
+	}
+	return va, nil
+}
+
+func (v *Volume) fetchTopClients(volumes []string, svms []string, metric string) ([]gjson.Result, error) {
+	var (
+		result []gjson.Result
+		err    error
+	)
+	if v.volumeInterface != nil {
+		return v.volumeInterface.fetchTopClients(volumes, svms, metric)
+	}
+	query := "api/storage/volumes/*/top-metrics/clients"
+	href := rest.NewHrefBuilder().
+		APIPath(query).
+		Fields([]string{"client_ip", "svm", "volume.name", metric}).
+		Filter([]string{"top_metric=" + metric, "volume=" + strings.Join(volumes, "|"), "svm=" + strings.Join(svms, "|")}).
```

Comment:
How likely is it that we hit the Apache request limit with this filter? Worth checking here?

Reply:
Let's do one of these:
1. In the current error handling block of `cmd/collectors/restperf/plugins/volumetopclients/volumetopclients.go` at line 393, check if the URL length is > 8KB and if so augment the returned error with that information
2. Change `cmd/collectors/restperf/plugins/volumetopclients/volumetopclients.go` InvokeRestCall to return the following at line 76
`logger.Error().Err(err).Str("href", href).Int("hrefLength", len(href)).Msg("Failed to fetch data")`